### PR TITLE
Rust CI

### PR DESCRIPTION
**Pull request by jackpot51**
Will merge this into impish when it looks good for testing with the impish build machine.

**Let @jackpot51 merge when approved to ensure Jenkins is transitioned at the same time**

### CODE REVIEW

File: scripts/pop-ci/src/git.rs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        async_std::process::Command::new("git")
+            .arg("-C").arg(&self.path())
+            .arg("fetch")
+            .arg("--prune")
+            .arg("--quiet")
+            .arg("--")
+            .arg(&remote.id())
+            .status()
+            .await
+            .and_then(check_status)
```
Comment (mmstick) [inline scripts/pop-ci/src/git.rs:81]: `async-std` will spawn a thread for each command being awaited. But we could use `async-pidfd` to await the status of a command without spawning a thread from a normal `std::process::Command`.

```rust
let child = command.spawn()?;
let pidfd = AsyncPidFd::from_pid(child.id())?;
check_status(pidfd.wait().await?.status())
```

File: scripts/pop-ci/src/cache.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        let path = self.path().join(name);
+        if path.exists() {
+            if force {
+                eprintln!("Cache::build: forcing rebuild of {}", path.display());
+                //TODO: rename before removing
+                if path.is_dir() {
+                    fs::remove_dir_all(&path)?;
+                } else {
+                    fs::remove_file(&path)?;
+                }
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply scripts/pop-ci/src/cache.rs:66]: It is important in the case more than one pop-ci instance is running at the same time. Alternatively, we could lock the cache directory so only one instance can run.

File: scripts/pop-ci/src/git.rs (hunk 1 above)
Comment (jackpot51) [reply scripts/pop-ci/src/git.rs:81]: > `async-std` will spawn a thread for each command being awaited. But we could use `async-pidfd` to await the status of a command without spawning a thread from a normal `std::process::Command`.
> 
> ```rust
> let child = command.spawn()?;
> let pidfd = AsyncPidFd::from_pid(child.id())?;
> check_status(pidfd.wait().await?.status())
> ```

That is a good idea, thanks

File: scripts/pop-ci/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,1024 @@
...
+        });
+    }
+
+    while let Some((name, res)) = futures.next().await {
+        if let Err(err) = res {
+            eprintln!("{}: failed to fetch git repo: {}", name, err);
+            process::exit(1);
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply scripts/pop-ci/src/main.rs:82]: Yep, limiting to maybe 64 would make me happy. Admittedly this code is some of the first parallel code I've done with async, so I think your rewrite here is a lot better

File: scripts/pop-ci/src/cache.rs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            } else {
+                return Ok((path, false));
+            }
+        }
+
+        let partial_path = self.path().join(format!("{}{}", partial_prefix, name));
+        if partial_path.exists() {
+            //TODO: should we automatically remove?
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply scripts/pop-ci/src/cache.rs:74]: This is technically done in Cache::new, with the retain function. It logs what it deletes. This ensures the cache has no dead items (including partial items) prior to using it.

File: scripts/pop-ci/src/main.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mmstick) [reply scripts/pop-ci/src/main.rs:82]: There's a few valid ways to schedule futures for parallel operations. This one uses the sub-scheduler approach, where multiple futures are polled concurrently from a single await point. Joining / zipping futures together performs the same kind of concurrent execution. Each future is cooperatively sharing the same thread the moment that they yield NotReady.

The other approach is spawning futures onto an executor. You get back a Task handle that returns the output of the future when you await the task, but the task is running in the background even if you aren't actively awaiting the task handle. That task could be running on a local executor from the same thread, an executor scheduling tasks on a background thread, or a runtime that's scheduling tasks on a thread pool.

And the executor you choose could be doing other tricks behind the scene, such as how smolscale will automatically interrupt and push long-running futures onto another thread, which can make sync code appear to seamlessly execute asynchronously.

**Pull request merged**